### PR TITLE
release: automate CHANGELOG generation

### PR DESCRIPTION
**Pull request by mccurdyc**
## Proposed Change(s)

* Add a make target for generating the CHANGELOG using the same approach as https://github.com/fastly/cli.
* Add a make target for generating a release description using the same approach as https://github.com/fastly/cli.
* Update CHANGELOG.md with some minor formatting tweaks that will align with
changelogs generated in the future.
   * [Rendered CHANGELOG.md](https://github.com/fastly/go-fastly/blob/9da007c0dbb12f09566452992a60641f8d23daf4/CHANGELOG.md)
* Update RELEASE docs
   * [Rendered RELEASE.md](https://github.com/fastly/go-fastly/blob/f024f130f0ca1ea3ed4e7734827835fe260003b5/RELEASE.md)

## Not Goal(s)

* Fully automating the release process.
  * Full automation can build on this change nicely.

## Design Decision(s)

* The name `CHANGELOG_HISTORY.md` was chosen because by default the [github-changelog-generator looks for a `HISTORY.md`](https://github.com/github-changelog-generator/github-changelog-generator#migrating-from-a-manual-changelog) which is nice for appending the automated changelog with the previously-manual changelog. However, this makes it impossible to make a changelog per release.
* [This approach](https://github.com/cli/cli/blob/trunk/script/changelog#L9-L13) was used to identify the previous tag, [unlike `egrep` in fastly/cli](https://github.com/fastly/cli/blob/master/Makefile#L4) in order to be more robust on different OSes.

## Validation

```
git clone git@github.com:mccurdyc/go-fastly.git /tmp/go-fastly &&
(
    cd /tmp/go-fastly
    git checkout mccurdyc/release
    make release-changelog
)
```

## Note(s)

* https://github.com/fastly/go-fastly/pull/200 uses this new process.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,6 +2,9 @@
 MKFILE_PATH := $(lastword $(MAKEFILE_LIST))
 CURRENT_DIR := $(patsubst %/,%,$(dir $(realpath $(MKFILE_PATH))))
 
+GITHUB_CHANGELOG_GENERATOR := $(shell command -v github_changelog_generator 2> /dev/null)
+PREVIOUS_SEMVER_TAG := $(shell git tag | sort -r --version-sort | grep -P '^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$$' | head -n2 | tail -n1)
```
Comment (mccurdyc) [inline Makefile:6]:
```suggestion
PREVIOUS_SEMVER_TAG := $(shell git tag | sort -r --version-sort | egrep '^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$$' | head -n2 | tail -n1)
```

[Note this difference between fastly/cli](https://github.com/fastly/cli/blob/master/Makefile#L4)

File: Makefile [hunk 2]
```diff
@@ -113,3 +116,41 @@ fix-fixtures:
...
+endif
+ifeq ($(SEMVER_TAG),)
+	$(error "You must set $$SEMVER_TAG to your desired release semver version.")
+endif
+	github_changelog_generator -u fastly -p go-fastly \
+		--future-release $(SEMVER_TAG) \
+		--no-pr-wo-labels \
+		--no-author \
+		--base CHANGELOG_HISTORY.md \
+		--since=v1.14.0 \
```
Comment (mccurdyc) [inline Makefile:137]: This is also different from the CLI.

**Comment by mccurdyc** [comment]
Force-pushed squashing into a single commit.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,223 +1,225 @@
...
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [Unreleased]
-
-## [1.14.0] - 2020-05-26
+## [v1.14.0] (2020-05-26)
 ### Added
-- ip: Support for accessing Fastly's IPv6 ranges ([#189](https://github.com/fastly/go-fastly/pull/189)).
+- ip: Support for accessing Fastly's IPv6 ranges ([v#189](https://github.com/fastly/go-fastly/pull/189)).
```
Comment (phamann) [inline CHANGELOG.md:12]: It seems there was a grep|sed error here as we now have a `v` character after every `[`. Similar comments for the rest of the file elided.

File: scripts/changelog.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	echo "You must set \$SEMVER_TAG to your desired release semver version."
+	exit 1
+fi
+
+github_changelog_generator -u fastly -p go-fastly \
+	--future-release $SEMVER_TAG \
+	--no-pr-wo-labels \
+	--no-author \
+	--base CHANGELOG_HISTORY.md \
+	--since "v1.14.0" \
```
Comment (phamann) [inline scripts/changelog.sh:23]: A test left in by accident? This shouldn't be hard coded.

Comment (phamann) [reply scripts/changelog.sh:23]: Oh wait, I think I get it, this is the first tag we started using the generator on?

Comment (mccurdyc) [reply scripts/changelog.sh:23]: Exactly!

File: CHANGELOG.md (hunk 3 above)
Comment (mccurdyc) [reply CHANGELOG.md]: https://github.com/fastly/go-fastly/pull/199/commits/d9b34a1ac761a9d8efcc122687b9cea7d7024ee5

File: CHANGELOG.md [hunk 5]
```diff
@@ -135,17 +135,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
-- dictionary_item: Support to create multiple items ([a162398](https://github.com/fastly/go-fastly/commit/a162398362d3a967cdc5518190344b16b5421060)).
+## [v0.4.3] (2019-05-24)
+### Added
+- waf: Support for OWASP ([v74c03ce](https://github.com/fastly/go-fastly/commit/74c03cec4549738bb1fc20fd881fbd9c750f6928), [vcf4b60f](https://github.com/fastly/go-fastly/commit/cf4b60fa9e61358753b9cafa4481adc5977d8432), [vcee4e1d](https://github.com/fastly/go-fastly/commit/cee4e1df35a27e507ef0809991fe8ffb462a94d4)).
```
Comment (phamann) [inline CHANGELOG.md:140]: Afraid we still have erroneous `v`'s in these commit references.

**Comment by mccurdyc** [comment]
Fixed the historical changelog and regenerated the changelog to ensure the diff didn't included erroneous 'v's.

File: scripts/tags.sh [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+set -e
+
+# credit: https://github.com/cli/cli/blob/trunk/script/changelog
+
+function previous_tag() {
+  current_tag="$(git describe --tags HEAD^ --abbrev=0)"
```
Comment (phamann) [inline scripts/tags.sh:7]: As we don't run this in CI on tag builds, I don't know if this is even required. I.e. all this function needs to do is get the previous tag which isn't a prerelease.

**Comment by mccurdyc** [comment]
Force-pushed squashing into a single commit.

File: scripts/tags.sh (hunk 6 above)
Comment (mccurdyc) [reply scripts/tags.sh:7]: Deleting :)

**Pull request merged**